crypto/src/key/hdkd/derivation_path.rs

```diff
@@ -0,0 +1,260 @@
+use core::fmt;
+use std::fmt::Formatter;
+use std::slice::Iter;
+use std::str::FromStr;
+
+const PREFIX: &str = "m";
+const SEPARATOR: &str = "/";
+const HARDENED_APOS: char = '\'';
+const HARDENED_H: char = 'h';
+const HARDENED_BIT: u32 = 0x80000000;
+
+/// BIP32-like child numbers. Currently we only support hardened derivations
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub struct ChildNumber(DerivationType);
+
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+enum DerivationType {
+    Hardened(u32),
+}
+
+impl ChildNumber {
+    /// Return a hardened child number
+    pub fn hardened(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            Ok(ChildNumber(DerivationType::Hardened(index)))
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a normal child number. Currently unsupported
+    fn normal(index: u32) -> Result<Self, DerivationError> {
+        if index & HARDENED_BIT == 0 {
+            // For the time being we don't support non-hardened derivations
+            Err(UnsupportedDerivationType)
+        } else {
+            Err(InvalidChildNumber(index))
+        }
+    }
+
+    /// Return a BIP32-like child number index that has the hardened bit set if needed
+    pub fn to_encoded_index(&self) -> u32 {
+        match self.0 {
+            DerivationType::Hardened(i) => i | HARDENED_BIT,
+        }
+    }
+
+    /// Returns true if this child number is hardened
+    pub fn is_hardened(&self) -> bool {
+        match self.0 {
+            DerivationType::Hardened(_) => true,
+        }
+    }
+}
+
+impl FromStr for ChildNumber {
+    type Err = DerivationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        if !s.is_ascii() {
+            return Err(InvalidDerivationPathFormat);
+        }
+        // Check if this child number is hardened
+        if let Some(s) = s.strip_suffix([HARDENED_APOS, HARDENED_H]) {
+            ChildNumber::hardened(s.parse().map_err(|_| InvalidChildNumberFormat)?)
+        } else {
+            ChildNumber::normal(s.parse().map_err(|_| InvalidChildNumberFormat)?)
+        }
+    }
+}
+
+impl fmt::Display for ChildNumber {
+    fn fmt(&self, f: &mut Formatter<'_>) -> fmt::Result {
+        match self.0 {
+            DerivationType::Hardened(index) => {
+                fmt::Display::fmt(&index, f)?;
+                let alt = f.alternate();
+                f.write_str(if alt { "h" } else { "'" })
+            }
+        }
+    }
+}
+
+/// BIP-32 compatible derivation path
+#[derive(Debug, Clone, PartialEq, Eq, Ord, PartialOrd)]
+pub struct DerivationPath(Vec<ChildNumber>);
+
+impl From<Vec<ChildNumber>> for DerivationPath {
+    fn from(path: Vec<ChildNumber>) -> Self {
+        DerivationPath(path)
+    }
+}
+
+impl<'a> IntoIterator for &'a DerivationPath {
+    type Item = &'a ChildNumber;
+    type IntoIter = Iter<'a, ChildNumber>;
```

Comment:
Minor: I'd prefer this to be a fully qualified name `std::slice::Iter` rather than importing `Iter` into the module namespace. Just to be more explicit, `Iter` is quite an ambiguous name.